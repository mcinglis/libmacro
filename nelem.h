
// Copyright 2015  Malcolm Inglis <http://minglis.id.au>
//
// This file is part of Libmacro.
//
// Libmacro is free software: you can redistribute it and/or modify it under
// the terms of the GNU Affero General Public License as published by the
// Free Software Foundation, either version 3 of the License, or (at your
// option) any later version.
//
// Libmacro is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public
// License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with Libmacro. If not, see <https://gnu.org/licenses/>.


#ifndef LIBMACRO_NELEM_H
#define LIBMACRO_NELEM_H


// Gives the number of elements in the array `XS`. Be very careful that
// you only call this with an *array* variable, and not a pointer-to-array.
// Note that this evaluates to a constant expression.
#define NELEM( ... ) \
    ( ( sizeof ( __VA_ARGS__ ) ) / ( sizeof ( ( __VA_ARGS__ )[ 0 ] ) ) )


#endif

