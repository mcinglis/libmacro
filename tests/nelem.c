
#include <stdio.h>
#include <libtypes/types.h>

#include "../nelem.h"


int main( void )
{
    int const xs[] = { 1, 2, 3, 4, 5, 6, 7, 8 };
    printf( "NELEM( xs ) = %zu\n", NELEM( xs ) );

    char const str[] = "testing string for NELEM!";
    printf( "NELEM( str ) = %zu\n", NELEM( str ) );

    typedef struct {
        char const * foo;
        size_t bar;
        ulong baz;
    } Widget;
    Widget const ws[] = { { .foo = "yup" }, { .bar = 123, .baz = 82374 },
                          { .baz = 0 }, { .foo = "huh?", .bar = 2193 } };
    printf( "NELEM( ws ) = %zu\n", NELEM( ws ) );
}

