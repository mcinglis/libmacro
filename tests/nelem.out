NELEM( xs ) = 8
NELEM( str ) = 26
NELEM( ws ) = 4
